File: db-query-backend/src/query/query.service.ts
import { Injectable, BadRequestException } from '@nestjs/common';
import { DataSource } from 'typeorm';
import { ConnectionManagerService } from '../connections/connection-manager.service';

@Injectable()
export class QueryService {
  constructor(private connectionManager: ConnectionManagerService) {}

  private async getDataSource(connectionId: string): Promise<DataSource> {
    // Get the pre-created DataSource from ConnectionManager
    const managedConnection = await this.connectionManager.getConnection(connectionId); // adjust method name
    if (!managedConnection || managedConnection.kind !== 'sql') {
      throw new BadRequestException('SQL connection not found');
    }

    const dataSource = managedConnection.ds;
    if (!dataSource.isInitialized) {
      await dataSource.initialize();
    }
    return dataSource;
  }

  private validateQuery(sql: string) {
    const forbidden = ['insert', 'update', 'delete', 'drop', 'alter', 'truncate'];

    const lowered = sql.toLowerCase();
    for (const word of forbidden) {
      if (lowered.includes(word)) {
        throw new BadRequestException('Only READ-ONLY queries are allowed');
      }
    }

    if (!lowered.trim().startsWith('select')) {
      throw new BadRequestException('Only SELECT queries are allowed');
    }
  }

  async executeQuery(sql: string, connectionId: string) {
    const dataSource = await this.getDataSource(connectionId);
    this.validateQuery(sql);

    const result = await dataSource.query(sql);
    return {
      rows: result,
      count: result.length,
    };
  }
}

File: db-query-backend/app/services/query_service.py
"""Port of query/query.service.ts"""
import re
from sqlalchemy import text

from app.services.connection_manager import connection_manager, MongoConnection

FORBIDDEN = re.compile(
    r"\b(insert|update|delete|drop|alter|truncate)\b", re.IGNORECASE
)


class QueryService:
    def execute(self, connection_id: str, query: str) -> list[dict]:
        conn = connection_manager.get(connection_id)

        if isinstance(conn, MongoConnection):
            raise ValueError("MongoDB connections are not supported for /query/execute")

        stripped = query.strip()
        if not stripped.lower().startswith("select"):
            raise ValueError("Only SELECT queries are allowed")
        if FORBIDDEN.search(stripped):
            raise ValueError("Query contains a forbidden keyword")

        with conn.engine.connect() as c:
            result = c.execute(text(stripped))
            columns = result.keys()
            return [dict(zip(columns, row)) for row in result.fetchall()]


query_service = QueryService()
